File: lib/seize/twitch.rb
require 'rest-client'

module Seize
  module Twitch
    BASE_URL = 'https://api.twitch.tv/kraken'
    HEADERS = {:accept => 'application/vnd.twitchtv.v3+json',
               :'client-id' => '2n7irufqjtyyayigyc4ubzq174axeex'}

    # TODO support games paramater
    def self.get_streams(limit: nil)
      streams_response = RestClient.get("#{BASE_URL}/streams", {
        :params => { :limit => limit }, :headers => HEADERS,
      })
      JSON.parse(streams_response)
    end

    def self.get_channel(channel_name)
      begin
        channel_response = RestClient.get(
          "#{BASE_URL}/streams/#{channel_name}",
          { :headers => HEADERS })
        JSON.parse(channel_response)
      rescue RestClient::NotFound
        nil
      end
    end
  end
end

File: src/twitch.js
import fetch from 'isomorphic-fetch';
import querystring from 'querystring';

const baseUrl = 'https://api.twitch.tv/kraken';
const headers = {
  accept: 'application/vnd.twitchtv.v5+json',
  'client-id': '2n7irufqjtyyayigyc4ubzq174axeex',
}

async function getStreams(limit = null, game = null) {
  let url = `${baseUrl}/streams?`;
  const queryParams = {};
  if (limit) {
    queryParams.limit = limit;
  }
  if (game) {
    queryParams.game = game;
  }
  url += querystring.stringify(queryParams);
  const { streams } = await fetchJsonWithHeaders(url);
  return streams;
}

async function searchStreams(query) {
  let url = `${baseUrl}/search/streams?`;
  url += querystring.stringify({ limit: 100, query });
  const { streams } = await fetchJsonWithHeaders(url);
  return streams;
}

async function fetchJsonWithHeaders(url, config = {}) {
  config.headers = Object.assign({}, config.headers, headers);
  const result = await fetch(url, config);
  return await result.json();
}

export default { getStreams, searchStreams };
